build-tests/api-documenter-test/etc/api-documenter-test.api.json

```diff
@@ -885,6 +911,7 @@
           "members": [
             {
               "kind": "MethodSignature",
+              "canonicalReference": "api-documenter-test!IDocInterface2#deprecatedExample:member(0)",
```

Comment:
@rbuckton I'm curious why `MethodSignature` and `PropertySignature` both end up as `:member` whereas `EnumMember` ends up as `:enummember`.

For an inexperienced person trying to learn these selector names, the feel a little inconsistent.  Is the idea to have as few selectors as possible so it's easy to memorize them (e.g. `:member`)?  Or is the idea that the selectors simply tell you want kind of thing it is (`:enummember`)?

Reply:
Enum Members are special in TypeScript as they have a type-specific meaning and can possibly be inlined (in the case of const enums). For example, `MyEnum.MyEnumMember` is both a static member of `MyEnum` as well as its own type.

The `:member` annotation is a combination of several TypeScript `SymbolFlags`, however they are mutually exclusive so it seemed safe to use the combined representation (called `SymbolFlags.ClassMember` in TypeScript, though that's a bit of a misnomer). Specifically, `SymbolFlags.Method`, `SymbolFlags.Property`, `SymbolFlags.GetAccessor`, and `SymbolFlags.SetAccessor`.

Reply:
In all fairness, it would probably be acceptable to drop `:enummember`, since they also shouldn't be able to collide, but then again `:variable` and `:function` would *also* be unnecessary using the same logic. Perhaps it would be better to use `:method`, `:property`, and `:accessor` instead of `:member` (though I don't know if I'd go so far as to split up `:accessor` into `:getaccessor` and `:setaccessor`)

Reply:
BTW I hadt assumed that `Property` and `PropertySignature` cannot conflict, but they actually can:

```ts
interface X {
  y: number; // X#y
}

class X {
  public y: number;  // X#y
}
```

Reply:
> Enum Members are special in TypeScript as they have a type-specific meaning and can possibly be inlined (in the case of const enums). For example, `MyEnum.MyEnumMember` is both a static member of `MyEnum` as well as its own type.

I'm not convinced that it's useful to distinguish "enum-member-value" and "enum-member-as-type" is a useful distinction in a documentation system.

Reply:
Here's everything that API Extractor deems worthy of referring to:
```ts
export const enum ApiItemKind {
  CallSignature = 'CallSignature',
  Class = 'Class',
  Constructor = 'Constructor',
  ConstructSignature = 'ConstructSignature',
  // EntryPoint = 'EntryPoint',
  Enum = 'Enum',
  EnumMember = 'EnumMember',
  Function = 'Function',
  IndexSignature = 'IndexSignature',
  Interface = 'Interface',
  Method = 'Method',
  MethodSignature = 'MethodSignature',
  // Model = 'Model',
  Namespace = 'Namespace',
  // Package = 'Package',
  Property = 'Property',
  PropertySignature = 'PropertySignature',
  TypeAlias = 'TypeAlias',
  Variable = 'Variable',
  // None = 'None'
}
```

If we want to collapse the list as small as possible, it might go as follows...

These have no name, so they must each get their own selector:
- `CallSignature` = `:call`
- `Constructor` = `:constructor`
- `ConstructSignature` = `:new`
- `IndexSignature` = `:index`

These things can never coexist at a given scope, and thus could be combined into a generalized name (let's call it `:other` for now):
- `Class`
- `Enum`
- `EnumMember`
- `Function`
- `Method`
- `Property`
- `TypeAlias`
- `Variable`

These can never coexist, and they only ever appear as part of an `interface`:
- `PropertySignature` = `:imember`
- `MethodSignature` = `:imember`

And finally these two promiscuous ones can merge with lots of things:
- `Namespace` = `:namespace`
- `Interface` = `:interface`

I'm concerned that such a partitioning is highly counterintuitive.  An everyday developer will really struggle to see any difference between `Method` and `MethodSignature`, let alone whatever `:other` is supposed to capture.

The language design wants `ApiItemKind` elements to be as orthogonal as possible, so their overlap is fractal and hard to summarize.

Reply:
(Theoretically `:type` and `:interface` could be combined, since they never coexist.)

Reply:
TSDoc's current notation tries to name the selector after the TypeScript keyword that you would use to declare the thing.  I think that's much easier for people to learn.  

That rule handles all of these:
- `Class` = `:class`
- `ConstructSignature` = `:new`
- `Constructor` = `:constructor`
- `Enum` = `:enum`
- `Function` = `:function`
- `Interface` = `:interface`
- `Namespace` = `:namespace`
- `TypeAlias` = `:type`
- `Variable` = `:var`

The rest have to be memorized, but as you pointed out a bunch are just "members" of a `class` or `enum` or `interface`:
- `CallSignature` = `:call` 
- `EnumMember` = `:member`  
- `IndexSignature` = `:index` 
- `MethodSignature` = `:imember`  
- `Method` = `:member`  
- `PropertySignature` = `:imember`  
- `Property` = `:member`

As I mentioned above, the `:imember` / `:member` conflict is somewhat artificial -- people don't normally make an interface with the same name as a class.  In practice the `:imember` will probably only ever be needed in the machine-generated canonical form.  We could allow humans to use `:member` instead, and if it's ambiguous they can write `X:interace#y:member` to avoid having to learn `:imember`.  

@rbuckton What do you think about this proposed change?  If you like it I will implement it as a TSDoc PR before merging this PR.

Reply:
I don't necessarily think going for a minimal representation is necessarily ideal. `:other` has no semantic meaning and could just be confusing. You may as well just be calling them `:1`, `:2, `:3`, etc. as it won't be easy to remember. Honestly, `:member` is probably fine for method/property/accessor. Remember that for the most part these meanings correspond to `SymbolFlags` in TypeScript, and the `ClassMember` bitmask covers all three. As I said before, `:enummember` is unique in that it can represent a type.

Also, there isn't much of a difference between `Property` and `PropertySignature` or `Method` and `MethodSignature`. The TypeScript compiler doesn't differentiate them when it comes to symbol meanings.

Finally, `:typealias` represents a TypeAlias. Not every `:typealias` is (or can be) an interface as they can represent some fairly complex types. `:type` is not intended to be used when creating a `DeclarationReference` for a declaration, but when creating a `DeclarationReference` for a complex type (union, intersection, object literal, mapped type, conditional type, etc) when generating documentation. It is distinct *precisely* so that it be used to synthesize deeply nested types and add reference and shouldn't be conflated with `:typealias` or `:interface`. For example:

```ts
// ref: package!Foo:interface
interface Foo<T> { 
}

// ref: package!Bar:interface
interface Bar {
}

// ref: package!Foo:function(1)
// return type: package!Foo~0:type
declare function Foo(): Foo<Bar>;
```
yaml:
```yaml
references:
- uid: package!Foo~0:type
  name: Foo<Bar>
  spec.typeScript:
  - name: Foo
    uid: package!Foo:interface
  - name: "<"
  - name: Bar
    uid: package!Bar:interface
  - name: ">"
```

Reply:
@rbuckton What does `package!Foo~0:type` mean?  I thought `~` was for referring to unexported members of a module/namespace?